### ISSUE TITLE
Save-Package Breaks When Feed Has Exactly 40 Entries With the Same Name

### ISSUE THREAD

**Issue by twinter-amosfivesix** [PowerShell/PowerShell#24657]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When using `Save-Package`, supplying a package name, required version, and using the NuGet provider, if the package feed does not support OData (like Azure Artifacts) and has exactly 40 entries for the requests package (regardless of version number), `Save-Package` fails with `Save-Package: Unable to install, multiple packages matched '<package name>'. Please specify an exact -Name and -RequiredVersion.`

If you use `Find-Package` under these circumstances, it returns *four* entries for the exact same package, when it should return only one.

# Why Does This Happen?

There is a bug in the OneGet NuGetProvider, `NuGetWebUtility.cs`, `SendRequest()`.

[The key code for this](https://github.com/OneGet/NuGetProvider/blob/6c486d0de4f4c550615467d2b1298e5fb0cdcd2b/Utility/NuGetWebUtility.cs#L81C1-L87C18).

This code is calling the NuGet `api/v2/FindPackagesById` endpoint. First it sends a single request, meant in part to see if the server support OData, which allows paging the results (using `skip` and `top` in the query string). Here I'm using a server that does *not* support OData (Azure Artifacts). This first request includes `skip` and `top` asking for the first 40 entries. Azure Artifacts doesn't support OData so it always returns all of the entries, ignoring `skip` and `top`. In our case it returns 40 entries, since that's how many package versions with this name are in the feed. The code looks at how many entries are returned. If it's less than 40 it (correctly) knows it's gotten all of the entries and stops asking for me. If it gets 40 or more entries, it assumes the server supports OData and so it sends out four more requests, each for a different page of data. Since our non-OData-supporting feed returns exactly 40 entries - since that's how many exist - the code *incorrectly* assumes the server support OData and so sends more requests. Each requests sends back the exact same entries, so the package we're looking for is returned multiple times. That's why `Find-Package` returns four entries for the same single package, and why `Save-Package` fails because it thinks there is more than one matching package.

Note that based on my understanding of the code, I would think it would return five entries for the package (one for the first requests, and one each for the subsequent four requests), but I only get four. 

# Workaround

No real one that I know of. Basically, make sure there aren't exactly 40 versions of your package in the feed. If you go down to 39 things work again or if you go up to 41 things work again.

### Expected behavior

```console
PS> Save-Package -Name PackageName -RequiredVersion 1.1.0 -Path .\path -ProviderName NuGet

Name          Version          Source           Summary
----          -------          ------           -------
PackageName   1.0.0            TheSource        Module description

PS> Find-Package -Name PackageName -RequiredVersion 1.0.0 -ProviderName NuGet 

Name          Version          Source           Summary
----          -------          ------           -------
PackageName   1.0.0            TheSource        Module description
```

### Actual behavior

```console
PS> Save-Package -Name PackageName -RequiredVersion 1.1.0 -Path .\path -ProviderName NuGet

Save-Package: Unable to install, multiple packages matched 'PackageName'. Please specify an exact -Name and -RequiredVersion.

PS> Find-Package -Name PackageName -RequiredVersion 1.0.0 -ProviderName NuGet 

Name          Version          Source           Summary
----          -------          ------           -------
PackageName   1.0.0            TheSource        Module description
PackageName   1.0.0            TheSource        Module description
PackageName   1.0.0            TheSource        Module description
PackageName   1.0.0            TheSource        Module description
```

### Error details

```console
(Sorry I don't have Get-Error output.)

I do have output from Find-Package -Verbose -Debug

A key thing to note here is that you do *not* get the PackagesReceived debug message ("'#' packages received in the last request."). That's because the number of entries received exactly matches the buffer size (line 83 in NuGetWebUtility.cs).

DEBUG: 00:00:00.0000032 Calling 'NuGet'::'ResolvePackageSources'.
DEBUG: 00:00:00.0000705 There are '2' registered sources in 'NuGet' provider.
DEBUG: 00:00:00.0001430 Yielding the package source: '«0»'. «NuGet» «Microsoft.PackageManagement.NuGetProvider.PackageSource»
DEBUG: 00:00:00.0001727 Yielding package source for TheSource at location https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
DEBUG: 00:00:00.0001946 Returning the call 'NuGet'::'ResolvePackageSources'.
VERBOSE: Suppressed Warning Unknown category for 'NuGet'::'GetDynamicOptions': 'Provider'
DEBUG: 00:00:00.1098635 Calling 'NuGet'::'ResolvePackageSources'.
DEBUG: 00:00:00.1099210 There are '2' registered sources in 'NuGet' provider.
DEBUG: 00:00:00.1099776 Yielding the package source: '«0»'. «NuGet» «Microsoft.PackageManagement.NuGetProvider.PackageSource»
DEBUG: 00:00:00.1099938 Yielding package source for TheSource at location https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
DEBUG: 00:00:00.1100105 Returning the call 'NuGet'::'ResolvePackageSources'.
DEBUG: 00:00:00.1102128 Calling SearchForPackages. Name='PackageName'
VERBOSE: Using the provider 'NuGet' for searching packages.
DEBUG: 00:00:00.1112423 PackageProvider::FindPackage with name PackageName
DEBUG: 00:00:00.1416527 Calling SearchForPackages After Select 1
DEBUG: 00:00:00.1427018 Calling 'NuGet'::'FindPackage' - name='PackageName', requiredVersion='1.0.0',minimumVersion='', maximumVersion='''.
DEBUG: 00:00:00.1427424 Iterating 'PackageName'.
DEBUG: 00:00:00.1431164 There are '2' registered sources in 'NuGet' provider.
DEBUG: 00:00:00.1432228 Calling 'NuGetRequest'::'GetPackageById', 'PackageName'.
DEBUG: 00:00:00.1434271 Calling 'NuGetPackageRepository'::'FindPackagesById', 'PackageName'.
DEBUG: 00:00:00.1434760 Calling 'NuGetPackageFeed2'::'FindPackage', 'PackageName'.
DEBUG: 00:00:00.1434902 Calling 'NuGetClient'::'FindPackage'.
VERBOSE: Searching repository 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'' for ''.
DEBUG: 00:00:00.1436210 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40'.
VERBOSE: Retry downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40' for '2' more times
VERBOSE: Retry downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40' for '1' more times
VERBOSE: Retry downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40' for '0' more times
DEBUG: 00:00:00.3315905 Calling 'GetCredsFromCredProvider' on https://pkgs.dev.azure.com:443/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=%7B0%7D&$top=%7B1%7D
DEBUG: 00:00:00.3318373 Calling credential provider installed at C:\Users\twinter/.nuget/plugins/netcore/CredentialProvider.Microsoft/CredentialProvider.Microsoft.dll
VERBOSE: [Verbose] [CredentialProvider]Running in stand-alone mode
VERBOSE: [Verbose] [CredentialProvider]Command-line v1.0.9+1c7de8e6b84724a3f145ae469036d5e57cbcbc61: C:\Users\twinter\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll -V verbose -U https://pkgs.dev.azure.com:443/my-organization/_packaging/TheSource/nuget/v2
VERBOSE: [Verbose] [CredentialProvider]Handling auth request, Uri: https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2, IsRetry: False, IsNonInteractive: False, CanShowDialog: True
VERBOSE: [Verbose] [CredentialProvider]URI: https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
VERBOSE: [Verbose] [CredentialProvider]VstsBuildTaskServiceEndpointCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials. Appropriate environment variable needs to be set.
VERBOSE: [Verbose] [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint.VstsBuildTaskServiceEndpointCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
VERBOSE: [Verbose] [CredentialProvider]VstsBuildTaskCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet. Appropriate environment variables must be set.
VERBOSE: [Verbose] [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTask.VstsBuildTaskCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
VERBOSE: [Verbose] [CredentialProvider]VstsCredentialProvider - Matched well-known Azure DevOps Service hostname: pkgs.dev.azure.com
VERBOSE: [Verbose] [CredentialProvider]Using NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider to try to get credentials for https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2.
VERBOSE: [Verbose] [CredentialProvider]IsRetry: False
VERBOSE: [Verbose] [CredentialProvider]Found cached SessionToken for https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2
DEBUG: 00:00:00.9758655 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40'.
DEBUG: 00:00:00.9795772 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40'.
DEBUG: 00:00:01.0977889 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=0&$top=40'.
DEBUG: 00:00:01.1035894 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=80&$top=40'.
DEBUG: 00:00:01.1035874 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=120&$top=40'.
DEBUG: 00:00:01.1035827 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=40&$top=40'.
DEBUG: 00:00:01.1036183 Downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=160&$top=40'.
DEBUG: 00:00:01.1683956 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=40&$top=40'.
DEBUG: 00:00:01.2372785 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=80&$top=40'.
DEBUG: 00:00:01.2380075 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=120&$top=40'.
DEBUG: 00:00:01.2409779 Completed downloading 'https://pkgs.dev.azure.com/my-organization/_packaging/TheSource/nuget/v2/FindPackagesById()?id='PackageName'&$skip=160&$top=40'.

Name          Version          Source           Summary
----          -------          ------           -------
PackageName   1.0.0            TheSource        Module description
PackageName   1.0.0            TheSource        Module description
PackageName   1.0.0            TheSource        Module description
VERBOSE: Total package yield:'4' for the specified package 'PackageName'.
DEBUG: 00:00:01.2669024 Completed iterating for 'PackageName'.
PackageName   1.0.0            TheSource        Module description
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.6
PSEdition                      Core
GitCommitId                    7.4.6
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
Note that the [`PackageManagement`](https://github.com/oneget/oneget) module (aka `OneGet`) isn't part of this repo, but, more importantly, it is _no longer in active development_.

**Comment by mklement0** [comment]
There is a third-party successor to `PackageManagement`, which the [`PackageManagement` "retirement" announcement](https://github.com/OneGet/oneget#announcement) links to, [`AnyPackage`](https://github.com/anypackage/anypackage):

  * Like `PackageManagement`, it uses a _provider model_ (though note that it does _not_ support providers created *for `PackageManagement`*).

  * Like `PackageManagement`, it [has a NuGet provider](https://www.anypackage.dev/docs/provider-catalog/nuget/about_NuGet_Provider) (personally, I have no experience with it).

**Comment by twinter-amosfivesix** [comment]
> Note that the [PackageManagement](https://github.com/oneget/oneget) module (aka OneGet) isn't part of this repo

Yeah I know. I've previously [logged a bug](https://github.com/OneGet/NuGetProvider/issues/68) there and got no response whatsoever, so I figured I'd try here. And at least I got a response. 😄 

> it is no longer in active development.

Yeah I didn't figure anyone was actually going to fix this. 😉 It's more for documenting it in case someone else runs across it.

> There is a third-party successor to PackageManagement,

I'm using `PackageManagement` just because it comes with PowerShell. Good to know about [AnyPackage](https://github.com/anypackage/anypackage) though. If this becomes a real problem for me, it looks like it should be usable for what I need.
